platform/linux-generic/include/odp_ring_st_internal.h

```diff
@@ -0,0 +1,109 @@
+/* Copyright (c) 2018, Linaro Limited
+ * All rights reserved.
+ *
+ * SPDX-License-Identifier:     BSD-3-Clause
+ */
+
+#ifndef ODP_RING_ST_INTERNAL_H_
+#define ODP_RING_ST_INTERNAL_H_
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#include <odp/api/hints.h>
+#include <odp_align_internal.h>
+
+/* Basic ring for single thread usage. Operations must be synchronized by using
+ * locks (or other means), when multiple threads use the same ring. */
+typedef struct {
+	uint32_t head;
+	uint32_t tail;
+	uint32_t mask;
+	uint32_t *data;
+
+} ring_st_t;
+
+/* Initialize ring. Ring size must be a power of two. */
+static inline void ring_st_init(ring_st_t *ring, uint32_t *data, uint32_t size)
+{
+	ring->head = 0;
+	ring->tail = 0;
+	ring->mask = size - 1;
+	ring->data = data;
+}
+
+/* Dequeue data from the ring head. Max_num is smaller than ring size.*/
+static inline uint32_t ring_st_deq_multi(ring_st_t *ring, uint32_t data[],
+					 uint32_t max_num)
+{
+	uint32_t head, tail, mask, idx;
+	uint32_t num, i;
+
+	head = ring->head;
+	tail = ring->tail;
+	mask = ring->mask;
+	num  = tail - head;
+
+	/* Empty */
+	if (num == 0)
+		return 0;
+
+	if (num > max_num)
+		num = max_num;
+
+	idx = head & mask;
+
+	for (i = 0; i < num; i++) {
+		data[i] = ring->data[idx];
+		idx     = (idx + 1) & mask;
+	}
+
+	ring->head = head + num;
```

Comment:
Note that neither `head` nor `tail` get masked when doing enqueues or dequeues, so they will traverse the entire 32-bit range of the variable over time. While the ring itself will never hold more than `size` entries, `head` and `tail` will wrap around. When you index into the ring you do so via a mask (`idx = head & mask;`), but calculating the number of elements in the ring doesn't do this, which is why `abs()` is needed.